src/downloader/worker.rs

```diff
@@ -0,0 +1,120 @@
+use super::{DownloadProgress, DownloadRequest};
+use crate::{error::DownloadError, Error};
+use reqwest::Client;
+use std::time::Duration;
+use tokio::{fs::File, io::AsyncWriteExt};
+
+pub(super) async fn download_thread(client: Client, mut req: DownloadRequest) {
+    fn should_retry(e: &Error) -> bool {
+        match e {
+            Error::Reqwest(network_err) => {
+                network_err.is_timeout()
+                    || network_err.is_connect()
+                    || network_err.is_request()
+                    || network_err
+                        .status()
+                        .map(|status_code| status_code.is_server_error())
+                        .unwrap_or(true)
+            }
+            Error::Download(DownloadError::Cancelled) | Error::Io(_) => false,
+            _ => false,
+        }
+    }
+
+    let mut last_error = None;
+    let max_attempts = req.config().max_retries();
+    for attempt in 0..=(max_attempts + 1) {
+        if attempt > max_attempts {
+            req.fail(Error::Download(DownloadError::RetriesExhausted {
+                last_error_msg: last_error
+                    .as_ref()
+                    .map(ToString::to_string)
+                    .unwrap_or_else(|| "Unknown Error".to_string()),
+            }));
+            return;
+        }
+
+        if attempt > 0 {
+            req.retry();
+            // Basic exponential backoff
+            let delay_ms = 1000 * 2u64.pow(attempt as u32 - 1);
+            let delay = Duration::from_millis(delay_ms);
+
+            tokio::select! {
+                _ = tokio::time::sleep(delay) => {},
+                _ = req.cancel.cancelled() => {
+                    req.cancel();
+                    return;
+                }
+            }
+        }
+
+        match download(client.clone(), &mut req).await {
+            Ok(file) => {
+                req.complete(file);
+                return;
+            }
+            Err(e) => {
+                if should_retry(&e) {
+                    last_error = Some(e);
+                    continue;
+                }
+
+                req.fail(e);
+                return;
+            }
+        }
+    }
+}
+
+async fn download(client: Client, req: &mut DownloadRequest) -> Result<File, Error> {
+    let mut response = client
+        .get(req.url().as_ref())
+        .send()
+        .await?
+        .error_for_status()?;
```

Comment:
We might want to use `config.user_agent()` to set the request’s User‑Agent header, here